File: distancedetectorjava/src/main/java/com/example/distancedetectorjava/CurrentLocationSendingService.java
package com.example.distancedetectorjava;

import android.app.Activity;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.provider.Settings;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.core.app.NotificationCompat;
import static com.example.distancedetectorjava.LocationConstants.CHANNEL_ID;
import static com.example.distancedetectorjava.LocationConstants.INTENT_KEY_NOTIFICATION_DETAIL;
import static com.example.distancedetectorjava.LocationConstants.INTENT_KEY_NOTIFICATION_TITLE;
import static com.example.distancedetectorjava.LocationConstants.INTENT_KEY_USER_ID;
import static com.example.distancedetectorjava.ServiceMessageHandler.broadcastingData;


public class CurrentLocationSendingService extends Service implements LocationListener {

   private Context mContext;
    // flag for GPS status
    boolean isGPSEnabled = false;

    // flag for network status
    boolean isNetworkEnabled = false;
    boolean canGetLocation = false;

    Location location;//Location
    double latitude;//Latitude
    double longitude;//Longitude
    String userId = "", title = "", details = "";

    // The minimum time between updates in milliseconds

    // Declaring a Location Manager
    protected LocationManager mlocationManager;


    public CurrentLocationSendingService() {
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        this.mContext = CurrentLocationSendingService.this;
        userId = intent.getStringExtra(INTENT_KEY_USER_ID);
        title = intent.getStringExtra(INTENT_KEY_NOTIFICATION_TITLE);
        details = intent.getStringExtra(INTENT_KEY_NOTIFICATION_DETAIL);
        //locationHandler.postDelayed(getLocationRunnable,0);
        getLocation();
        createNotificationChannel();


        //do heavy work on a background thread


        //stopSelf();

        return START_NOT_STICKY;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        //getLocation();
        createNotificationChannel();
    }

//    Handler locationHandler = new Handler();
//    Runnable getLocationRunnable =new Runnable() {
//        @Override
//        public void run() {
//            Location location = getLocation();
//            broadcastingData(mContext, location.getLatitude(), location.getLongitude(), userId, LocationConstants.EVENT_FROM_GPS);
//            //Toast.makeText(mContext,""+location.getLatitude(),Toast.LENGTH_LONG).show();
//            locationHandler.postDelayed(getLocationRunnable,2000);
//        }
//    };

    public Location getLocation() {
        try {
            mlocationManager = (LocationManager) mContext.getSystemService(LOCATION_SERVICE);
            isGPSEnabled = mlocationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
            isNetworkEnabled = mlocationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
            if (!isGPSEnabled && !isNetworkEnabled) {
                // no network provider is enabled
            } else {
                this.canGetLocation = true;
                // First get location from Network Provider
                if (isNetworkEnabled) {
                    //noinspection MissingPermission
                    mlocationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, LocationConstants.MIN_TIME_BW_UPDATES, LocationConstants.MIN_DISTANCE_BW_UPDATES, this);
                    Log.d("Network", "Network");
                    if (mlocationManager != null) {
                        //noinspection MissingPermission
                        location = mlocationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
                        if (location != null) {
                            latitude = location.getLatitude();
                            longitude = location.getLongitude();

                        }
                    }
                }

                if (isGPSEnabled) {
                    if (location == null) {
                        //noinspection MissingPermission
                        mlocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, LocationConstants.MIN_TIME_BW_UPDATES, LocationConstants.MIN_DISTANCE_BW_UPDATES, this);
                        Log.d("GPS Enabled", "GPS Enabled");
                        if (mlocationManager != null) {
                            //noinspection MissingPermission
                            location = mlocationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
                            if (location != null) {
                                latitude = location.getLatitude();
                                longitude = location.getLongitude();
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return location;
    }

    public void stopUsingGPS() {
        if (mlocationManager != null) {
            mlocationManager.removeUpdates(CurrentLocationSendingService.this);
        }
    }

    public double getLatitude() {
        if (location != null) {
            latitude = location.getLatitude();
        }


        return latitude;
    }

    public double getLongitude() {
        if (location != null) {
            longitude = location.getLongitude();
        }

        return longitude;
    }

    public boolean canGetLocation() {
        return this.canGetLocation;
    }


    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }


    @Override
    public void onLocationChanged(Location location) {
        this.location = location;

        double latitude = getLatitude();
        double longitude = getLongitude();

        broadcastingData(mContext, latitude, longitude, userId, LocationConstants.EVENT_FROM_GPS);
    }

    @Override
    public void onStatusChanged(String s, int i, Bundle bundle) {

    }

    @Override
    public void onProviderEnabled(String s) {

    }

    @Override
    public void onProviderDisabled(String s) {

    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel serviceChannel = new NotificationChannel(
                    CHANNEL_ID,
                    "Foreground Service Channel",
                    NotificationManager.IMPORTANCE_DEFAULT
            );

            NotificationManager manager = getSystemService(NotificationManager.class);
            manager.createNotificationChannel(serviceChannel);
            Intent notificationIntent = new Intent(this, Context.class);
            PendingIntent pendingIntent = PendingIntent.getActivity(this,
                    0, notificationIntent, 0);

            Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                    .setContentTitle(title)
                    .setContentText(details)
                    .setSmallIcon(R.mipmap.ic_launcher)
                    .setContentIntent(pendingIntent)
                    .build();

            startForeground(1, notification);
        }
    }

    public static boolean isLocationEnabled(final Activity activity) {

        try {
            LocationManager mlocationManager = (LocationManager) activity.getSystemService(LOCATION_SERVICE);
            boolean isGPSEnabled = mlocationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
            boolean isNetworkEnabled = mlocationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);

            if (!isGPSEnabled && !isNetworkEnabled) {
                AlertDialog.Builder alertDialog = new AlertDialog.Builder(activity);

                alertDialog.setTitle("GPS is settings");

                alertDialog.setMessage("GPS is not enabled. Do you want to go to settings menu?");

                alertDialog.setPositiveButton("Settings", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                        activity.startActivity(intent);
                    }
                });

                alertDialog.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                });

                alertDialog.show();

                return false;
            } else {
                return true;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }


    }

    @Override
    public void onDestroy() {
        stopUsingGPS();
        //locationHandler.removeCallbacks(getLocationRunnable);
        super.onDestroy();
    }
}

File: distancedetector/src/main/java/com/example/distancedectector/CurrentLocationSendingService.kt
package com.example.distancedectector

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.util.Log
import android.widget.Toast
import androidx.core.app.NotificationCompat
import com.example.distancedectector.LocationConstants.Companion.CHANNEL_ID
import com.example.distancedectector.LocationConstants.Companion.INTENT_KEY_NOTIFICATION_DETAIL
import com.example.distancedectector.LocationConstants.Companion.INTENT_KEY_NOTIFICATION_TITLE
import com.example.distancedectector.ServiceMessageHandler.Companion.broadcastingData

class CurrentLocationSendingService : Service(),LocationListener{


    private lateinit var mContext: Context
    private var isGPSEnabled = false

    private var isNetworkEnabled = false
    private var canGetLocation = false

    private var location: Location? = null
    private var lat: Double = 0.0
    private var lng: Double = 0.0
    private var title:String? = ""
    private var details:String? = ""
    private var locationHandler = Handler()
    private var mlocationManager: LocationManager? = null


    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        this.mContext = this@CurrentLocationSendingService
        title = intent!!.getStringExtra(INTENT_KEY_NOTIFICATION_TITLE)
        details = intent.getStringExtra(INTENT_KEY_NOTIFICATION_DETAIL)
        //locationHandler.postDelayed(getLocationRunnable,0)
        getLocation()
        createNotificationChannel()

        return START_NOT_STICKY
    }

    /*private var getLocationRunnable: Runnable = object : Runnable {
        override fun run() {
            val location = getLocation()
            if (location != null) {
                Toast.makeText(mContext,""+location.latitude,Toast.LENGTH_SHORT).show()
                broadcastingData(mContext, location.latitude,location.longitude)
            }else{
                broadcastingData(mContext, 0.0,0.0)
            }
            locationHandler.postDelayed(this, 1000)
        }
    }*/

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
    }


    fun getLocation(): Location? {
        try {
            mlocationManager = mContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            isGPSEnabled = mlocationManager!!.isProviderEnabled(LocationManager.GPS_PROVIDER)
            isNetworkEnabled = mlocationManager!!.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
            if (!isGPSEnabled && !isNetworkEnabled) {
                // no network provider is enabled
            } else {
                this.canGetLocation = true
                // First get location from Network Provider
                if (isNetworkEnabled) {

                    mlocationManager!!.requestLocationUpdates(
                        LocationManager.NETWORK_PROVIDER,
                        LocationConstants.MIN_TIME_BW_UPDATES,
                        LocationConstants.MIN_DISTANCE_BW_UPDATES.toFloat(),
                        this
                    )
                    Log.d("Network", "Network")
                    if (mlocationManager != null) {

                        location = mlocationManager!!.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
                        if (location != null) {
                            lat = location!!.latitude
                            lng = location!!.longitude

                        }
                    }
                }

                if (isGPSEnabled) {
                    if (location == null) {

                        mlocationManager!!.requestLocationUpdates(
                            LocationManager.GPS_PROVIDER,
                            LocationConstants.MIN_TIME_BW_UPDATES,
                            LocationConstants.MIN_DISTANCE_BW_UPDATES.toFloat(),
                            this
                        )
                        Log.d("GPS Enabled", "GPS Enabled")
                        if (mlocationManager != null) {

                            location = mlocationManager!!.getLastKnownLocation(LocationManager.GPS_PROVIDER)
                            if (location != null) {
                                lat = location!!.latitude
                                lng = location!!.longitude
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return location
    }

    fun stopUsingGPS() {
        if (mlocationManager != null) {
            mlocationManager!!.removeUpdates(this@CurrentLocationSendingService)
        }
    }

    fun getLatitude(): Double {
        if (location != null) {
            lat = location!!.latitude
        }


        return lat
    }

    fun getLongitude(): Double {
        if (location != null) {
            lng = location!!.longitude
        }

        return lng
    }


    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val serviceChannel = NotificationChannel(
                CHANNEL_ID,
                "Foreground Service Channel",
                NotificationManager.IMPORTANCE_DEFAULT
            )

            val manager = getSystemService(NotificationManager::class.java)
            manager!!.createNotificationChannel(serviceChannel)
            val notificationIntent = Intent(this, Context::class.java)
            val pendingIntent = PendingIntent.getActivity(
                this,
                0, notificationIntent, 0
            )

            val notification = NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle(title)
                .setContentText(details)
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentIntent(pendingIntent)
                .build()

            startForeground(1, notification)
        }
    }



    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onLocationChanged(location: Location?) {
        this.location = location

        val latitude = getLatitude()
        val longitude = getLongitude()

        //Toast.makeText(this,""+latitude +longitude,Toast.LENGTH_SHORT).show()
        broadcastingData(mContext, latitude, longitude)
    }

    override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {
    }

    override fun onProviderEnabled(provider: String?) {
    }

    override fun onProviderDisabled(provider: String?) {
    }


    override fun onDestroy() {
        super.onDestroy()
        //locationHandler.removeCallbacks(getLocationRunnable)
        stopUsingGPS()
    }




}